src/kaleidoscope/Device.h

```diff
@@ -0,0 +1,477 @@
+ * Copyright (C) 2017, 2018, 2019  Keyboard.io, Inc
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file kaleidoscope/Hardware.h
+ * Base class for Kaleidoscope hardware libraries.
+ */
+
+#pragma once
+
+#include "kaleidoscope/MatrixAddr.h"
+#include "kaleidoscope_internal/deprecations.h"
+#include "kaleidoscope/macro_helpers.h"
+
+#ifndef CRGB
+#error cRGB and CRGB *must* be defined before including this header!
+#endif
+
+#define _DEPRECATED_MESSAGE_HARDWARE_RESETDEVICE                       \
+  "`KeyboardHardware.resetDevice()` is deprecated, please use " __NL__ \
+  "`KeyboardHardware.rebootBootloader()` instead."
+
+/* All hardware libraries must define the following macros:
+ * HARDWARE_IMPLEMENTATION - the name of your public object conforming to
+ *   the 'class Hardware' interface below.
+ * CRGB(r,g,b) - explained below
+ * cRGB, a structure with at least three members: r, g, and b -
+ * compilation will fail otherwise.
+ *
+ * Despite its name, the members do not need to be in the order r g b -- most
+ * likely they will be in an order that is convenient for the hardware. So
+ * initializing a cRGB with a struct literal will give surprising results for any
+ * colors where r, g, and b do not have the same value. Each Hardware library
+ * defines a CRGB(r,g,b) macro which returns a literal cRGB with the given values.
+ */
+
+namespace kaleidoscope {
+
+/** Kaleidoscope Hardware base class.
+ * Essential methods all hardware libraries must implement.
+ */
+
+template<typename _DeviceDescription>
+class Device {
+ private:
+  class DummyIO {
+   public:
+    DummyIO() {}
+  };
+
+  static DummyIO dummy_io_;
+
+ public:
+
+  typedef typename _DeviceDescription::KeyScanner KeyScanner;
+  typedef typename _DeviceDescription::KeyScannerDescription::KeyAddr KeyAddr;
+  typedef typename _DeviceDescription::LEDs LEDs;
+  typedef typename _DeviceDescription::MCU MCU;
+  typedef typename _DeviceDescription::BootLoader BootLoader;
+  typedef typename _DeviceDescription::Storage Storage;
+
+  static constexpr uint8_t matrix_rows = _DeviceDescription::KeyScannerDescription::matrix_rows;
+  static constexpr uint8_t matrix_columns = _DeviceDescription::KeyScannerDescription::matrix_columns;
+  static constexpr typename _DeviceDescription::LEDs::LedCountType led_count = _DeviceDescription::LEDs::led_count;
+  static constexpr auto &focus_interface = dummy_io_;
```

Comment:
This could be replaced by
```cpp
static constexpr auto &focus_interface() -> decltype(dummy_io_) {
   return dummy_io_;
}
```
it would be in sync with the rest of the controllers for which access methods are provided.

Reply:
At some point, I had a reason for going with a member variable instead of an accessor... I even wrote it down somewhere. But it looks like I rebased that reason out of existence... *sigh*

I'll rework it to be an accessor, for the sake of consistency.